tx-pool/src/component/pool_map.rs

```diff
@@ -0,0 +1,588 @@
+        let mut removed_ids = vec![id.to_owned()];
+        let mut removed = vec![];
+        removed_ids.extend(self.calc_descendants(id));
+
+        // update links state for remove, so that we won't update_descendants_index_key in remove_entry
+        for id in &removed_ids {
+            self.remove_entry_links(id);
+        }
+
+        for id in removed_ids {
+            if let Some(entry) = self.remove_entry(&id) {
+                removed.push(entry);
+            }
+        }
+        removed
+    }
+
+    pub(crate) fn resolve_conflict_header_dep(
+        &mut self,
+        headers: &HashSet<Byte32>,
+    ) -> Vec<ConflictEntry> {
+        let mut conflicts = Vec::new();
+
+        // invalid header deps
+        let mut ids = Vec::new();
+        for (tx_id, deps) in self.edges.header_deps.iter() {
+            for hash in deps {
+                if headers.contains(hash) {
+                    ids.push((hash.clone(), tx_id.clone()));
+                    break;
+                }
+            }
+        }
+
+        for (blk_hash, id) in ids {
+            let entries = self.remove_entry_and_descendants(&id);
+            for entry in entries {
+                let reject = Reject::Resolve(OutPointError::InvalidHeader(blk_hash.to_owned()));
+                conflicts.push((entry, reject));
+            }
+        }
+        conflicts
+    }
+
+    pub(crate) fn resolve_conflict(&mut self, tx: &TransactionView) -> Vec<ConflictEntry> {
+        let inputs = tx.input_pts_iter();
+        let mut conflicts = Vec::new();
+
+        for i in inputs {
+            if let Some(id) = self.edges.remove_input(&i) {
+                let entries = self.remove_entry_and_descendants(&id);
+                if !entries.is_empty() {
+                    let reject = Reject::Resolve(OutPointError::Dead(i.clone()));
+                    let rejects = std::iter::repeat(reject).take(entries.len());
+                    conflicts.extend(entries.into_iter().zip(rejects));
+                }
+            }
+
+            // deps consumed
+            if let Some(x) = self.edges.remove_deps(&i) {
+                for id in x {
+                    let entries = self.remove_entry_and_descendants(&id);
+                    if !entries.is_empty() {
+                        let reject = Reject::Resolve(OutPointError::Dead(i.clone()));
+                        let rejects = std::iter::repeat(reject).take(entries.len());
+                        conflicts.extend(entries.into_iter().zip(rejects));
+                    }
+                }
+            }
+        }
+
+        conflicts
+    }
+
+    // fill proposal txs
+    pub(crate) fn fill_proposals(
+        &self,
+        limit: usize,
+        exclusion: &HashSet<ProposalShortId>,
+        proposals: &mut HashSet<ProposalShortId>,
+        status: &Status,
+    ) {
+        for entry in self.entries.get_by_status(status) {
+            if proposals.len() == limit {
+                break;
+            }
```

Comment:
Should we break the for loop only when `proposals.len() == limit`?
I think `proposals.len() >= limit` is more safe?

Reply:
it's added one-by-one, so it makes no difference? I think it's also good to change to `>=`.